File: BOJ/2162.cpp
#include <cstdio>
#include <algorithm>
#include <cstring>
#include <vector>
#include <queue>
#include <map>
#include <iostream>
#include <string>
#include <math.h>
#include <set>
#include <list>
#include <climits>
#include <string.h>
#include <deque>
#include <functional>
#include <stack>
using namespace std;
typedef long long ll;
#define INF 1000000000
#define mod 1000000009
typedef pair<ll,ll> P;
typedef pair<pair<int, int>, int> PP;
typedef pair<pair<ll, ll>, pair<ll, ll>> PPP;
int gox[4] = {0,1, -1,0 };
int goy[4] = {1,0, 0,-1 };
int n;
vector<PPP> v;
int parent[3005];
int Size[3005];
ll ccw(P A, P B, P C) {
	ll x1 = A.first, y1 = A.second, x2 = B.first, y2 = B.second, x3 = C.first, y3 = C.second;
	return (x1*y2 + x2 * y3 + x3 * y1) - (x2*y1 + x3 * y2 + x1 * y3);
}
bool cross(P A,P B,P C,P D) {
	ll x1 = A.first, y1 = A.second, x2 = B.first, y2 = B.second, x3 = C.first, y3 = C.second, x4 = D.first, y4 = D.second;
	if (x1 == x2 && x2 == x3 && x3 == x4 && x4 == x1) {
		ll maxy = max(y1, y2);
		ll miny = min(y3, y4);
		if (maxy < miny) return false;
		maxy = max(y3, y4);
		miny = min(y1, y2);
		if (maxy < miny) return false;
	}
	else {
		ll maxx = max(x1, x2);
		ll minx = min(x3, x4);
		if (maxx < minx) return false;
		maxx = max(x3, x4);
		minx = min(x1, x2);
		if (maxx < minx) return false;
	}
	return true;
}
int find(int u) {
	if (parent[u] == u) return u;
	return parent[u] = find(parent[u]);
}
void uni(int u, int v) {
	u = find(u), v = find(v);
	if (u == v) return;
	if (Size[u] < Size[v]) {
		Size[v] += Size[u];
		Size[u] = 0;
		parent[u] = v;
	}
	else {
		Size[u] += Size[v];
		Size[v] = 0;
		parent[v] = u;
	}
}
int main() {
	scanf("%d", &n);
	for (int i = 1; i <= n; i++) parent[i] = i;
	for (int i = 0; i <= n; i++) Size[i] = 1;
	for (int i = 1; i <= n; i++) {
		ll x1, y1, x2, y2;
		scanf("%lld%lld%lld%lld", &x1, &y1, &x2, &y2);
		v.push_back({{ x1,y1 }, { x2,y2 }});
	}
	for (int i = 0; i < v.size(); i++) {
		for (int j = i + 1; j < v.size(); j++) {
			PPP here = v[i], next = v[j];
			ll ccw1 = ccw(here.first,here.second,next.first);
			ll ccw2 = ccw(here.first, here.second, next.second);
			ll ccw3 = ccw(next.first, next.second, here.first);
			ll ccw4 = ccw(next.first, next.second, here.second);
			if (ccw1*ccw2 > 0 || ccw3 * ccw4 > 0) continue;
			else {
				if (ccw1*ccw2 == 0 && ccw3*ccw4 == 0 && !cross(here.first, here.second, next.first, next.second)) continue;
				if (find(i) != find(j)) uni(i, j);
			}
		}
	}
	int cnt = 0,Max=0;
	for (int i = 0; i < n; i++) {
		if (Size[i] != 0) {
			cnt++;
			Max = max(Size[i], Max);
		}
	}
	printf("%d\n%d\n", cnt, Max);
	return 0;
}

File: BOJ(Java 공부용)/2162.java
import java.io.*;
import java.util.ArrayList;
import java.util.List;
import java.util.StringTokenizer;

import static java.lang.Long.max;
import static java.lang.Long.min;

public class Main {
    static BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        List<PP> v = new ArrayList<>();
        int[] parent = new int[n];
        int[] size = new int[n];
        for (int i = 0; i < n; i++) {
            parent[i] = i;
            size[i] = 1;
        }
        for (int i = 0; i < n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine());
            long x1 = Long.parseLong(st.nextToken()), y1 = Long.parseLong(st.nextToken()), x2 = Long.parseLong(st.nextToken()), y2 = Long.parseLong(st.nextToken());
            v.add(new PP(x1, y1, x2, y2));
        }
        for (int i = 0; i < v.size(); i++) {
            for (int j = i + 1; j < v.size(); j++) {
                PP here = v.get(i);
                PP next = v.get(j);
                long ccw1 = ccw(here.x1, here.y1, here.x2, here.y2, next.x1, next.y1);
                long ccw2 = ccw(here.x1, here.y1, here.x2, here.y2, next.x2, next.y2);
                long ccw3 = ccw(next.x1, next.y1, next.x2, next.y2, here.x1, here.y1);
                long ccw4 = ccw(next.x1, next.y1, next.x2, next.y2, here.x2, here.y2);
                if (ccw1 * ccw2 > 0 || ccw3 * ccw4 > 0) continue;
                else {
                    if (ccw1 * ccw2 == 0 && ccw3 * ccw4 == 0 && !cross(here.x1, here.y1, here.x2, here.y2, next.x1, next.y1, next.x2, next.y2))
                        continue;
                    if (find(i, parent) != find(j, parent)) union(i, j, parent, size);
                }
            }
        }
        int cnt = 0, max = 0;
        for (int i = 0; i < n; i++) {
            if (size[i] != 0) {
                cnt++;
                if (max < size[i]) max = size[i];
            }
        }
        bw.write(cnt + "\n" + max + "\n");
        bw.flush();
        bw.close();
    }

    private static void union(int u, int v, int[] parent, int[] Size) {
        u = find(u, parent);
        v = find(v, parent);
        if (u == v) return;
        if (Size[u] < Size[v]) {
            Size[v] += Size[u];
            Size[u] = 0;
            parent[u] = v;
        } else {
            Size[u] += Size[v];
            Size[v] = 0;
            parent[v] = u;
        }
    }

    private static int find(int u, int[] parent) {
        if (u == parent[u]) return u;
        return parent[u] = find(parent[u], parent);
    }

    private static boolean cross(long x1, long y1, long x2, long y2, long x3, long y3, long x4, long y4) {
        if (x1 == x2 && x2 == x3 && x3 == x4) {
            return !check(y1, y2, y3, y4);
        } else {
            return !check(x1, x2, x3, x4);
        }
    }

    private static boolean check(long x1, long x2, long x3, long x4) {
        long maxx = max(x1, x2);
        long minx = min(x3, x4);
        if (maxx < minx) return true;
        maxx = max(x3, x4);
        minx = min(x1, x2);
        return maxx < minx;
    }

    private static long ccw(long x1, long y1, long x2, long y2, long x3, long y3) {
        return (x1 * y2 + x2 * y3 + x3 * y1) - (x2 * y1 + x3 * y2 + x1 * y3);
    }
}

class PP {
    long x1, y1, x2, y2;

    PP(long x1, long y1, long x2, long y2) {
        this.x1 = x1;
        this.y1 = y1;
        this.x2 = x2;
        this.y2 = y2;
    }
}

class P {
    long x;
    long y;

    P(long x, long y) {
        this.x = x;
        this.y = y;
    }
}